### PR TITLE
[FLINK-3230] Add producer for Amazon Kinesis Streams (flink -> kinesis)

### PR DESCRIPTION
**Pull request by rmetzger**
I've tested this connector against two kinesis streams. Since Kinesis is a hosted service, I can not implement a full integration test that runs with each test.
I'll run further tests once we've added a Kinesis consumer as well.

### CODE REVIEW

File: flink-streaming-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/FlinkKinesisProducer.java [hunk 1]
```diff
@@ -0,0 +1,272 @@
...
+			}
+		});
+	}
+
+	public FlinkKinesisProducer(String region, String accessKey, String secretKey, KinesisSerializationSchema<OUT> schema) {
+		this.region = Objects.requireNonNull(region);
+		this.accessKey = Objects.requireNonNull(accessKey);
+		this.secretKey = Objects.requireNonNull(secretKey);
+		ClosureCleaner.ensureSerializable(Objects.requireNonNull(schema));
+		this.schema = schema;
```
Comment (uce) [inline flink-streaming-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/FlinkKinesisProducer.java]: no `null` check on purpose?

Comment (rmetzger) [inline flink-streaming-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/FlinkKinesisProducer.java]: There is a null check on the schema as well ;)

File: flink-streaming-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/FlinkKinesisProducer.java [hunk 2]
```diff
@@ -0,0 +1,272 @@
...
+		super.open(parameters);
+
+		KinesisProducerConfiguration config = new KinesisProducerConfiguration();
+		config.setRegion(this.region);
+		config.setCredentialsProvider(new StaticCredentialsProvider(new BasicAWSCredentials(this.accessKey, this.secretKey)));
+		producer = new KinesisProducer(config);
+		callback = new FutureCallback<UserRecordResult>() {
+			@Override
+			public void onSuccess(UserRecordResult result) {
+				if(!result.isSuccessful()) {
```
Comment (uce) [inline flink-streaming-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/FlinkKinesisProducer.java]: if style space missing (also in other places)

File: flink-streaming-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/KinesisSerializationSchema.java [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+	 * Serialize the given element into a ByteBuffer
+	 *
+	 * @param element The element to serialize
+	 * @return Serialized representation of the element
+	 */
+	ByteBuffer serialize(T element);
+
+	/**
+	 * Optional method to determine the target stream based on the element.
+	 * Return "null" to use the default stream
```
Comment (uce) [inline flink-streaming-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/KinesisSerializationSchema.java]: Do you mean to return the String `"null"` or `<code>null</code>`?

File: flink-streaming-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/examples/ProduceIntoKinesis.java [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+
+	public static void main(String[] args) throws Exception {
+		ParameterTool pt = ParameterTool.fromArgs(args);
+
+		StreamExecutionEnvironment see = StreamExecutionEnvironment.getExecutionEnvironment();
+		see.setParallelism(1);
+
+		DataStream<String> simpleStringStream = see.addSource(new EventsGenerator());
+
+		FlinkKinesisProducer<String> kinesis = new FlinkKinesisProducer<>(pt.getRequired("region"),
```
Comment (uce) [inline flink-streaming-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/examples/ProduceIntoKinesis.java]: I would have each argument on its own line

File: flink-streaming-connectors/flink-connector-twitter/pom.xml [hunk 5]
```diff
@@ -35,6 +35,11 @@ under the License.
 
 	<packaging>jar</packaging>
 
+	<properties>
+		<hbc-core.version>2.2.0</hbc-core.version>
```
Comment (uce) [inline flink-streaming-connectors/flink-connector-twitter/pom.xml]: Why do we have this as a property now? Is it overwritten somewhere?

Comment (rmetzger) [inline flink-streaming-connectors/flink-connector-twitter/pom.xml]: I wanted to allow users passing a custom property using `-Dhbc-core.version=a.b.c`. I didn't change the version, just introduced the prop.

File: flink-streaming-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/FlinkKinesisProducer.java [hunk 6]
```diff
@@ -0,0 +1,272 @@
...
+
+
+	/* Our Kinesis instance for each parallel Flink sink */
+	private transient KinesisProducer producer;
+
+	/* Callback handling failures */
+	private transient FutureCallback<UserRecordResult> callback;
+
+	/* Field for async exception */
+	private transient Throwable thrownException;
```
Comment (uce) [inline flink-streaming-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/FlinkKinesisProducer.java]: this should be volatile, if the callback is executed by a different thread

File: flink-streaming-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/FlinkKinesisProducer.java (hunk 1 above)
Comment (uce) [inline flink-streaming-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/FlinkKinesisProducer.java]: Ah didn't see it in the argument to the ClosureCleaner

File: flink-streaming-connectors/flink-connector-twitter/pom.xml (hunk 5 above)
Comment (uce) [inline flink-streaming-connectors/flink-connector-twitter/pom.xml]: OK

File: flink-streaming-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/FlinkKinesisProducer.java [hunk 7]
```diff
@@ -0,0 +1,272 @@
...
+		Futures.addCallback(cb, callback);
+	}
+
+	@Override
+	public void close() throws Exception {
+		super.close();
+		KinesisProducer kp = this.producer;
+		this.producer = null;
+		if(kp != null) {
+			LOG.info("Closing producer. Flushing outstanding {} records", kp.getOutstandingRecordsCount());
```
Comment (uce) [inline flink-streaming-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/FlinkKinesisProducer.java]: `"Flushing {} outstanding records"`?

File: flink-streaming-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/FlinkKinesisProducer.java [hunk 8]
```diff
@@ -0,0 +1,272 @@
...
+		if(customPartitioner != null) {
+			partition = customPartitioner.getPartitionId(value);
+			explicitHashkey = customPartitioner.getExplicitHashKey(value);
+		}
+
+		if(stream == null) {
+			if(failOnError) {
+				throw new RuntimeException("No target stream set");
+			} else {
+				LOG.warn("No target stream set. Skipping record");
```
Comment (uce) [inline flink-streaming-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/FlinkKinesisProducer.java]: If I don't provide a default stream and the schema returns no target partition, this can never be resolved, no? If this is true, we should fail here?

File: docs/apis/streaming/connectors/kinesis.md [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+
+Unless required by applicable law or agreed to in writing,
+software distributed under the License is distributed on an
+"AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+KIND, either express or implied.  See the License for the
+specific language governing permissions and limitations
+under the License.
+-->
+
+The Kinesis connector allows to produce data into an [Amazon AWS Kinesis Stream](http://aws.amazon.com/kinesis/streams/).
```
Comment (uce) [inline docs/apis/streaming/connectors/kinesis.md]: Should we add that there are no exactly-once guarantees?

File: flink-streaming-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/FlinkKinesisProducer.java [hunk 10]
```diff
@@ -0,0 +1,272 @@
...
+	@Override
+	public void close() throws Exception {
+		super.close();
+		KinesisProducer kp = this.producer;
+		this.producer = null;
+		if(kp != null) {
+			LOG.info("Closing producer. Flushing outstanding {} records", kp.getOutstandingRecordsCount());
+			// try to flush all outstanding records
+			while (kp.getOutstandingRecordsCount() > 0) {
+				kp.flush();
```
Comment (zentol) [inline flink-streaming-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/FlinkKinesisProducer.java]: can flush() throw an exception? if so they should be catched to guarantee that kp.destroy() is called.

File: flink-streaming-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/examples/ProduceIntoKinesis.java [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+	}
+
+	public static class EventsGenerator implements SourceFunction<String> {
+		private boolean running = true;
+
+		@Override
+		public void run(SourceContext<String> ctx) throws Exception {
+			long seq = 0;
+			while(running) {
+
```
Comment (zentol) [inline flink-streaming-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/examples/ProduceIntoKinesis.java]: this empty line seems odd.

File: flink-streaming-connectors/flink-connector-twitter/pom.xml (hunk 5 above)
Comment (zentol) [inline flink-streaming-connectors/flink-connector-twitter/pom.xml]: how is this change related to the Kinesis connector?

**Comment by uce** [comment]
Great addition to have a producer to Kinesis. The changes look good overall, I had some minor inline comments.

The main concern is that we can only test this with a real Kinesis setup, which makes it fragile for future changes etc. Did you consider trying to mock the `KinesisProducer` to test that the expected calls are issued to it?

PS: The title of the JIRA issue and this PR is confusing (the last `... into Flink` part)

File: flink-streaming-connectors/flink-connector-twitter/pom.xml (hunk 5 above)
Comment (rmetzger) [inline flink-streaming-connectors/flink-connector-twitter/pom.xml]: Its not related at all. I just came across it and fixed it.

File: flink-streaming-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/FlinkKinesisProducer.java (hunk 8 above)
Comment (rmetzger) [inline flink-streaming-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/FlinkKinesisProducer.java]: the schema can set a target stream on each record. So it can happen that a user didn't specify a default stream but the schema returns a value stream name for 99% of the data.

File: flink-streaming-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/FlinkKinesisProducer.java (hunk 10 above)
Comment (rmetzger) [inline flink-streaming-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/FlinkKinesisProducer.java]: No, flush() doesn't throw exceptions.

File: flink-streaming-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/examples/ProduceIntoKinesis.java (hunk 11 above)
Comment (rmetzger) [inline flink-streaming-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/examples/ProduceIntoKinesis.java]: Its invisible code ;)

**Comment by rmetzger** [comment]
Thank you @zentol and @uce for the review.
I hope I addressed all your concerns.

File: flink-streaming-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/FlinkKinesisProducer.java (hunk 8 above)
Comment (uce) [inline flink-streaming-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/FlinkKinesisProducer.java]: Yeah, sure. I'm just saying that it is easy to shoot yourself in the foot, if you forget to set a default stream and don't use a fancy schema. Then, the user will just wonder what is happening and the log will be flooded. What do you think?

For your example case, the warning is OK with the `failOnError` configuration.

Comment (rmetzger) [inline flink-streaming-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/FlinkKinesisProducer.java]: Okay, I see. I'll set the "fail on error" behavior to "true" by default.

**Comment by rmetzger** [comment]
Merging ...
